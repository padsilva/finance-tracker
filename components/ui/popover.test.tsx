import { render } from "@testing-library/react";

import {
  Popover,
  PopoverTrigger,
  PopoverContent,
  PopoverAnchor,
} from "./popover";

// Mock portal to avoid ReactDOM.createPortal issues in tests
jest.mock("@radix-ui/react-popover", () => ({
  Root: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="popover-root">{children}</div>
  ),
  Trigger: ({ children }: { children: React.ReactNode }) => (
    <button data-testid="popover-trigger">{children}</button>
  ),
  Anchor: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="popover-anchor">{children}</div>
  ),
  Portal: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="popover-portal">{children}</div>
  ),
  Content: ({
    children,
    className,
    align,
    sideOffset,
  }: {
    children: React.ReactNode;
    className?: string;
    align?: string;
    sideOffset?: number;
  }) => (
    <div
      data-testid="popover-content"
      className={className}
      data-align={align}
      data-side-offset={sideOffset}
    >
      {children}
    </div>
  ),
}));

describe("Popover Components", () => {
  it("should render basic popover correctly", () => {
    const { container } = render(
      <Popover>
        <PopoverTrigger>Open Popover</PopoverTrigger>
        <PopoverContent>
          <div>Basic Popover Content</div>
        </PopoverContent>
      </Popover>,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render with custom alignment", () => {
    const { container } = render(
      <Popover>
        <PopoverTrigger>Align Test</PopoverTrigger>
        <PopoverContent align="start">Start Aligned Content</PopoverContent>
        <PopoverContent align="center">Center Aligned Content</PopoverContent>
        <PopoverContent align="end">End Aligned Content</PopoverContent>
      </Popover>,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render with custom offset", () => {
    const { container } = render(
      <Popover>
        <PopoverTrigger>Offset Test</PopoverTrigger>
        <PopoverContent sideOffset={8}>
          Content with Custom Offset
        </PopoverContent>
      </Popover>,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render with custom className", () => {
    const { container } = render(
      <Popover>
        <PopoverTrigger className="custom-trigger">
          Styled Trigger
        </PopoverTrigger>
        <PopoverContent className="custom-content">
          Custom Styled Content
        </PopoverContent>
      </Popover>,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render with anchor", () => {
    const { container } = render(
      <Popover>
        <PopoverAnchor>
          <div>Anchor Element</div>
        </PopoverAnchor>
        <PopoverTrigger>Open</PopoverTrigger>
        <PopoverContent>Anchored Content</PopoverContent>
      </Popover>,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render with different animation states", () => {
    const { container } = render(
      <Popover>
        <PopoverTrigger>Animation Test</PopoverTrigger>
        <PopoverContent data-state="open">Open State Content</PopoverContent>
        <PopoverContent data-state="closed">
          Closed State Content
        </PopoverContent>
      </Popover>,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render with different side positions", () => {
    const { container } = render(
      <Popover>
        <PopoverTrigger>Side Test</PopoverTrigger>
        <PopoverContent data-side="top">Top Content</PopoverContent>
        <PopoverContent data-side="right">Right Content</PopoverContent>
        <PopoverContent data-side="bottom">Bottom Content</PopoverContent>
        <PopoverContent data-side="left">Left Content</PopoverContent>
      </Popover>,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render with complex content", () => {
    const { container } = render(
      <Popover>
        <PopoverTrigger>Open Complex Popover</PopoverTrigger>
        <PopoverContent>
          <div className="grid gap-4">
            <div className="space-y-2">
              <h4 className="font-medium leading-none">Dimensions</h4>
              <p className="text-sm text-muted-foreground">
                Set the dimensions for the layer.
              </p>
            </div>
            <div className="grid gap-2">
              <div className="grid grid-cols-3 items-center gap-4">
                <label htmlFor="width">Width</label>
                <input
                  id="width"
                  className="col-span-2 h-8"
                  type="number"
                  defaultValue="100%"
                />
              </div>
              <div className="grid grid-cols-3 items-center gap-4">
                <label htmlFor="height">Height</label>
                <input
                  id="height"
                  className="col-span-2 h-8"
                  type="number"
                  defaultValue="25px"
                />
              </div>
            </div>
          </div>
        </PopoverContent>
      </Popover>,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render with nested content", () => {
    const { container } = render(
      <Popover>
        <PopoverTrigger>Open Nested</PopoverTrigger>
        <PopoverContent>
          <Popover>
            <PopoverTrigger>Nested Trigger</PopoverTrigger>
            <PopoverContent>Nested Content</PopoverContent>
          </Popover>
        </PopoverContent>
      </Popover>,
    );

    expect(container).toMatchSnapshot();
  });
});
