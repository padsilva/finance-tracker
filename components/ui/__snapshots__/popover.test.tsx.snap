// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover Components should render basic popover correctly 1`] = `
<div>
  <div
    data-testid="popover-root"
  >
    <button
      data-testid="popover-trigger"
    >
      Open Popover
    </button>
    <div
      data-testid="popover-portal"
    >
      <div
        class="z-50 w-72 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
        data-align="center"
        data-side-offset="4"
        data-testid="popover-content"
      >
        <div>
          Basic Popover Content
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Popover Components should render with anchor 1`] = `
<div>
  <div
    data-testid="popover-root"
  >
    <div
      data-testid="popover-anchor"
    >
      <div>
        Anchor Element
      </div>
    </div>
    <button
      data-testid="popover-trigger"
    >
      Open
    </button>
    <div
      data-testid="popover-portal"
    >
      <div
        class="z-50 w-72 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
        data-align="center"
        data-side-offset="4"
        data-testid="popover-content"
      >
        Anchored Content
      </div>
    </div>
  </div>
</div>
`;

exports[`Popover Components should render with complex content 1`] = `
<div>
  <div
    data-testid="popover-root"
  >
    <button
      data-testid="popover-trigger"
    >
      Open Complex Popover
    </button>
    <div
      data-testid="popover-portal"
    >
      <div
        class="z-50 w-72 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
        data-align="center"
        data-side-offset="4"
        data-testid="popover-content"
      >
        <div
          class="grid gap-4"
        >
          <div
            class="space-y-2"
          >
            <h4
              class="font-medium leading-none"
            >
              Dimensions
            </h4>
            <p
              class="text-sm text-muted-foreground"
            >
              Set the dimensions for the layer.
            </p>
          </div>
          <div
            class="grid gap-2"
          >
            <div
              class="grid grid-cols-3 items-center gap-4"
            >
              <label
                for="width"
              >
                Width
              </label>
              <input
                class="col-span-2 h-8"
                id="width"
                type="number"
                value="100%"
              />
            </div>
            <div
              class="grid grid-cols-3 items-center gap-4"
            >
              <label
                for="height"
              >
                Height
              </label>
              <input
                class="col-span-2 h-8"
                id="height"
                type="number"
                value="25px"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Popover Components should render with custom alignment 1`] = `
<div>
  <div
    data-testid="popover-root"
  >
    <button
      data-testid="popover-trigger"
    >
      Align Test
    </button>
    <div
      data-testid="popover-portal"
    >
      <div
        class="z-50 w-72 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
        data-align="start"
        data-side-offset="4"
        data-testid="popover-content"
      >
        Start Aligned Content
      </div>
    </div>
    <div
      data-testid="popover-portal"
    >
      <div
        class="z-50 w-72 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
        data-align="center"
        data-side-offset="4"
        data-testid="popover-content"
      >
        Center Aligned Content
      </div>
    </div>
    <div
      data-testid="popover-portal"
    >
      <div
        class="z-50 w-72 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
        data-align="end"
        data-side-offset="4"
        data-testid="popover-content"
      >
        End Aligned Content
      </div>
    </div>
  </div>
</div>
`;

exports[`Popover Components should render with custom className 1`] = `
<div>
  <div
    data-testid="popover-root"
  >
    <button
      data-testid="popover-trigger"
    >
      Styled Trigger
    </button>
    <div
      data-testid="popover-portal"
    >
      <div
        class="z-50 w-72 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2 custom-content"
        data-align="center"
        data-side-offset="4"
        data-testid="popover-content"
      >
        Custom Styled Content
      </div>
    </div>
  </div>
</div>
`;

exports[`Popover Components should render with custom offset 1`] = `
<div>
  <div
    data-testid="popover-root"
  >
    <button
      data-testid="popover-trigger"
    >
      Offset Test
    </button>
    <div
      data-testid="popover-portal"
    >
      <div
        class="z-50 w-72 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
        data-align="center"
        data-side-offset="8"
        data-testid="popover-content"
      >
        Content with Custom Offset
      </div>
    </div>
  </div>
</div>
`;

exports[`Popover Components should render with different animation states 1`] = `
<div>
  <div
    data-testid="popover-root"
  >
    <button
      data-testid="popover-trigger"
    >
      Animation Test
    </button>
    <div
      data-testid="popover-portal"
    >
      <div
        class="z-50 w-72 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
        data-align="center"
        data-side-offset="4"
        data-testid="popover-content"
      >
        Open State Content
      </div>
    </div>
    <div
      data-testid="popover-portal"
    >
      <div
        class="z-50 w-72 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
        data-align="center"
        data-side-offset="4"
        data-testid="popover-content"
      >
        Closed State Content
      </div>
    </div>
  </div>
</div>
`;

exports[`Popover Components should render with different side positions 1`] = `
<div>
  <div
    data-testid="popover-root"
  >
    <button
      data-testid="popover-trigger"
    >
      Side Test
    </button>
    <div
      data-testid="popover-portal"
    >
      <div
        class="z-50 w-72 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
        data-align="center"
        data-side-offset="4"
        data-testid="popover-content"
      >
        Top Content
      </div>
    </div>
    <div
      data-testid="popover-portal"
    >
      <div
        class="z-50 w-72 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
        data-align="center"
        data-side-offset="4"
        data-testid="popover-content"
      >
        Right Content
      </div>
    </div>
    <div
      data-testid="popover-portal"
    >
      <div
        class="z-50 w-72 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
        data-align="center"
        data-side-offset="4"
        data-testid="popover-content"
      >
        Bottom Content
      </div>
    </div>
    <div
      data-testid="popover-portal"
    >
      <div
        class="z-50 w-72 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
        data-align="center"
        data-side-offset="4"
        data-testid="popover-content"
      >
        Left Content
      </div>
    </div>
  </div>
</div>
`;

exports[`Popover Components should render with nested content 1`] = `
<div>
  <div
    data-testid="popover-root"
  >
    <button
      data-testid="popover-trigger"
    >
      Open Nested
    </button>
    <div
      data-testid="popover-portal"
    >
      <div
        class="z-50 w-72 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
        data-align="center"
        data-side-offset="4"
        data-testid="popover-content"
      >
        <div
          data-testid="popover-root"
        >
          <button
            data-testid="popover-trigger"
          >
            Nested Trigger
          </button>
          <div
            data-testid="popover-portal"
          >
            <div
              class="z-50 w-72 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
              data-align="center"
              data-side-offset="4"
              data-testid="popover-content"
            >
              Nested Content
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
